File: Benchmarks/Core/Collections/YTB.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Runtime.CompilerServices;
using Benchmarks.Core.Entity;

namespace Benchmarks.Core.Collections
{
    /// <summary>
    /// Arreglo dinamico de alto rendimiento usado por el motor.
    /// <para>High-performance growable array used by the engine.</para>
    /// </summary>
    public class YTB<T>
    {
        /// <summary>
        /// Tamanos predeterminados para recalcular el tamano de YTB.
        /// </summary>
        private readonly int[] _predetSizesOfYTB = [500, 1000, 5000, 10_000, 50_000, 100_000, 500_000, 1_000_000, 5_000_000, 10_000_000, 20_000_000, 30_000_000, 40_000_000, 50_000_000];

        /// <summary>
        /// Almacena el indice actual en el arreglo de tamanos predeterminados.
        /// </summary>
        private int predetCurrentSizeIndex;

        /// <summary>
        /// Propiedad para acceder al indice actual en el arreglo de tamanos predeterminados.
        /// </summary>
        public int Capacity { get => _predetSizesOfYTB[predetCurrentSizeIndex]; private set => predetCurrentSizeIndex = value < 14 ? value : predetCurrentSizeIndex; }

        /// <summary>
        /// Iterable interno que almacena los elementos.
        /// </summary>
        internal T[] _arr { get; set; }

        public IEnumerable<T> _ytb => _arr.Take(Count);

        /// <summary>
        /// Numero de elementos en el arreglo.
        /// </summary>
        public int Count { get; private set; }

        /// <summary>
        /// Inicializa una nueva instancia de la clase YTB con capacidad inicial predeterminada.
        /// </summary>
        public YTB()
        {
            predetCurrentSizeIndex = 0;
            _arr = new T[Capacity];
            Count = 0;
        }

        /// <summary>
        /// Agrega un elemento al final del arreglo, redimensionando si es necesario.
        /// </summary>
        [MethodImpl(MethodImplOptions.AggressiveInlining)]
        public int Add(T item)
        {
            if (Count >= Capacity)
            {
                predetCurrentSizeIndex++;
                if (predetCurrentSizeIndex >= _predetSizesOfYTB.Length)
                    throw new System.IndexOutOfRangeException("YTB has reached its maximum capacity.");

                T[] newArray = new T[Capacity];
                Array.Copy(_arr, newArray, Count);
                _arr = newArray;
            }
            _arr[Count] = item;
            return ((Count++));
        }

        /// <summary>
        /// Agrega un elemento al final del arreglo, redimensionando si es necesario.
        /// </summary>
        [MethodImpl(MethodImplOptions.AggressiveInlining)]
        public int Add(T item, out int Index)
        {
            if (Count >= Capacity)
            {
                predetCurrentSizeIndex++;
                if (predetCurrentSizeIndex >= _predetSizesOfYTB.Length)
                    throw new System.IndexOutOfRangeException("YTB has reached its maximum capacity.");

                T[] newArray = new T[Capacity];
                Array.Copy(_arr, newArray, Count);
                _arr = newArray;
            }
            _arr[Count] = item;
            Index = ((Count++));
            return Index;
        }

        /// <summary>
        /// Remueve el elemento en la posicion indicada y desplaza los elementos posteriores.
        /// </summary>
        public bool RemoveAt(int index)
        {
            if (index < 0 || index >= Count)
                return false;
            int shiftCount = Count - index - 1;
            if (shiftCount > 0)
            {
                Array.Copy(_arr, index + 1, _arr, index, shiftCount);
            }
            _arr[--Count] = default!;
            return true;
        }

        /// <summary>
        /// Elimina la primera ocurrencia de un elemento especificado del YTB.
        /// </summary>
        public bool Remove(T item)
        {
            int index = Array.IndexOf(_arr, item, 0, Count);
            if (index < 0)
                return false;
            return RemoveAt(index);
        }

        /// <summary>
        /// Obtiene una referencia del objeto en la posicion indicada.
        /// </summary>
        public ref T this[int index]
        {
            [MethodImpl(MethodImplOptions.AggressiveInlining)]
            get
            {
                if (index < 0 || index >= Count)
                    throw new ArgumentOutOfRangeException(nameof(index), "Index is out of range.");
                return ref _arr[index];
            }
        }

        /// <summary>
        /// Obtiene una referencia del objeto asociado a la entidad indicada.
        /// </summary>
        public ref T this[Yotsuba entity]
        {
            [MethodImpl(MethodImplOptions.AggressiveInlining)]
            get
            {
                if (entity.Id < 0 || entity.Id >= Count)
                    throw new ArgumentOutOfRangeException(nameof(entity.Id), "Index is out of range.");
                return ref _arr[entity.Id];
            }
        }

        /// <summary>
        /// Establece un valor en una posicion especifica usando indice uint.
        /// </summary>
        public T this[uint index]
        {
            [MethodImpl(MethodImplOptions.AggressiveInlining)]
            set
            {
                if (index < 0 || index >= Count)
                    throw new ArgumentOutOfRangeException(nameof(index), "Index is out of range.");
                _arr[index] = value;
            }
        }

        public T[] ToArray()
        {
            T[] result = new T[Count];
            Array.Copy(_arr, result, Count);
            return result;
        }

        public List<T> ToList()
        {
            List<T> List = new List<T>(Count);
            List.AddRange(_arr.AsSpan(0, Count));
            return List;
        }

        public ReadOnlySpan<T> AsReadOnlySpan()
        {
            return _arr.AsSpan(0, Count);
        }

        public Span<T> AsSpan()
        {
            return _arr.AsSpan(0, Count);
        }

        public void Clear()
        {
            Count = 0;
            Capacity = 0;
            Array.Clear(_arr, 0, Capacity);
        }

        T DefaultValue = default(T);
        public ref T Find(Predicate<T> predicate)
        {
            int index = Array.FindIndex(_arr, 0, Count, predicate);
            if (index == -1)
            {
                return ref DefaultValue;
            }
            return ref _arr[index];
        }
    }
}

File: Benchmarks_FSharp/Core/Collections/YTB.fs
namespace Benchmarks.Core.Collections

open System
open System.Collections.Generic
open System.Linq
open System.Runtime.CompilerServices
open Benchmarks.Core.Entity

type YTB<'T>() =
    let _predetSizesOfYTB = [| 500; 1000; 5000; 10_000; 50_000; 100_000; 500_000; 1_000_000; 5_000_000; 10_000_000; 20_000_000; 30_000_000; 40_000_000; 50_000_000 |]
    let mutable predetCurrentSizeIndex = 0
    let mutable _arr: 'T array = Array.zeroCreate _predetSizesOfYTB.[0]
    let mutable _count = 0
    let mutable _defaultValue: 'T = Unchecked.defaultof<'T>

    member _.Capacity
        with get() = _predetSizesOfYTB.[predetCurrentSizeIndex]

    member internal _.InternalArray
        with get() = _arr

    member _._ytb = _arr.Take(_count)

    member _.Count
        with get() = _count

    [<MethodImpl(MethodImplOptions.AggressiveInlining)>]
    member _.Add(item: 'T) =
        if _count >= _predetSizesOfYTB.[predetCurrentSizeIndex] then
            predetCurrentSizeIndex <- predetCurrentSizeIndex + 1
            if predetCurrentSizeIndex >= _predetSizesOfYTB.Length then
                raise (IndexOutOfRangeException("YTB has reached its maximum capacity."))
            let newArray = Array.zeroCreate _predetSizesOfYTB.[predetCurrentSizeIndex]
            Array.Copy(_arr, newArray, _count)
            _arr <- newArray
        _arr.[_count] <- item
        let index = _count
        _count <- _count + 1
        index

    [<MethodImpl(MethodImplOptions.AggressiveInlining)>]
    member _.Add(item: 'T, index: outref<int>) =
        if _count >= _predetSizesOfYTB.[predetCurrentSizeIndex] then
            predetCurrentSizeIndex <- predetCurrentSizeIndex + 1
            if predetCurrentSizeIndex >= _predetSizesOfYTB.Length then
                raise (IndexOutOfRangeException("YTB has reached its maximum capacity."))
            let newArray = Array.zeroCreate _predetSizesOfYTB.[predetCurrentSizeIndex]
            Array.Copy(_arr, newArray, _count)
            _arr <- newArray
        _arr.[_count] <- item
        let idx = _count
        _count <- _count + 1
        index <- idx
        idx

    member _.RemoveAt(index: int) =
        if index < 0 || index >= _count then
            false
        else
            let shiftCount = _count - index - 1
            if shiftCount > 0 then
                Array.Copy(_arr, index + 1, _arr, index, shiftCount)
            _count <- _count - 1
            _arr.[_count] <- Unchecked.defaultof<'T>
            true

    member this.Remove(item: 'T) =
        let index = Array.IndexOf(_arr, item, 0, _count)
        if index < 0 then false
        else this.RemoveAt(index)

    member _.Item
        with get(index: int) : 'T byref =
            if index < 0 || index >= _count then
                raise (ArgumentOutOfRangeException(nameof index, "Index is out of range."))
            &_arr.[index]

    member _.ItemByEntity
        with get(entity: Yotsuba) : 'T byref =
            if entity.Id < 0 || entity.Id >= _count then
                raise (ArgumentOutOfRangeException("entity.Id", "Index is out of range."))
            &_arr.[entity.Id]

    member _.SetByUint(index: uint, value: 'T) =
        if int index >= _count then
            raise (ArgumentOutOfRangeException(nameof index, "Index is out of range."))
        _arr.[int index] <- value

    member _.ToArray() =
        let result = Array.zeroCreate _count
        Array.Copy(_arr, result, _count)
        result

    member _.ToList() =
        let list = List<'T>(_count)
        list.AddRange(_arr.AsSpan(0, _count))
        list

    member _.AsReadOnlySpan() : ReadOnlySpan<'T> =
        _arr.AsSpan(0, _count) |> ReadOnlySpan.op_Implicit

    member _.AsSpan() : Span<'T> =
        _arr.AsSpan(0, _count)

    member _.Clear() =
        _count <- 0
        predetCurrentSizeIndex <- 0
        Array.Clear(_arr, 0, _predetSizesOfYTB.[0])

    member _.Find(predicate: Predicate<'T>) : 'T byref =
        let index = Array.FindIndex(_arr, 0, _count, predicate)
        if index = -1 then
            &_defaultValue
        else
            &_arr.[index]
